### PR TITLE
Feat/ipfs corenode integration

### PR DESCRIPTION
**Pull request by ianopolous**
This isn't strictly ready to merge yet (I need to add a utility to generate keypairs), but I'd like feedback on the rest. 

This actually starts using the ipfs core node. This meant making it easy to bootstrap a new peergos network with a peergos user and pki keypair for tests and ram only mode.

### CODE REVIEW

File: src/peergos/server/Start.java [hunk 1]
```diff
@@ -86,9 +90,87 @@
             Collections.emptyList()
     );
 
+    public static final Command BOOTSTRAP = new Command("bootstrap",
```
Comment (kevodwyer) [inline src/peergos/server/Start.java]: maybe expand on documentation to highlight that the peergos.password should be specified

File: src/peergos/server/corenode/IpfsCoreNode.java [hunk 2]
```diff
@@ -26,6 +26,7 @@
 
     public static final PublicSigningKey PEERGOS_IDENTITY_KEY = PublicSigningKey.fromString("ggFYIE7uD1ViM9KfiA1w69n774/jk6hERINN3xACPyabWiBp");
     public static final PublicKeyHash PEERGOS_IDENTITY_KEY_HASH = PublicKeyHash.fromString("zdpuAvZynWLuyvovJwa34bj24M7cspt5M8seFfrFLrPWDGFDW");
+    public static final PublicSigningKey PKI_PUBLIC_KEY = null; // Fill in once generated
```
Comment (kevodwyer) [inline src/peergos/server/corenode/IpfsCoreNode.java]: not sure i understand...

File: src/peergos/server/Start.java (hunk 1 above)
Comment (cboddy) [reply src/peergos/server/Start.java]: +1

File: src/peergos/server/Start.java [hunk 3]
```diff
@@ -86,9 +90,87 @@
...
+            "Bootstrap a new peergos network",
+            args -> {
+                try {
+                    Crypto crypto = Crypto.initJava();
+                    // setup peergos user and pki keys
+                    String testpassword = args.getArg("peergos.password", "testpassword");
```
Comment (cboddy) [inline src/peergos/server/Start.java]: sad face as discussed

File: src/peergos/server/Start.java [hunk 4]
```diff
@@ -86,9 +90,87 @@
...
+                            crypto.random, crypto.signer, crypto.boxer, UserGenerationAlgorithm.getDefault()).get().getUser();
+                    dht.putSigningKey(peergosIdentityKeys.secretSigningKey.signatureOnly(
+                            pkiKeys.publicSigningKey.serialize()),
+                            peergosPublicHash,
+                            pkiKeys.publicSigningKey).get();
+
+                    Files.write(Paths.get("test.pki.secret.key"), pkiKeys.secretSigningKey.toCbor().toByteArray());
+                    Files.write(Paths.get("test.pki.public.key"), pkiKeys.publicSigningKey.toCbor().toByteArray());
+                    args.setIfAbsent("peergos.identity.hash", peergosPublicHash.toString());
+                    args.setIfAbsent("pki.secret.key", "test.pki.secret.key");
```
[the remark this answers is not in the document]
Comment (cboddy) [reply src/peergos/server/Start.java]: as discussed  these will be  removed

File: src/peergos/server/Start.java [hunk 5]
```diff
@@ -86,9 +90,87 @@
...
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    System.exit(1);
+                }
+            },
+            Collections.emptyList()
+    );
+
+    public static final Command POSTSTRAP = new Command("poststrap",
+            "The final step of bootstrapping a new peergos network, which must be run after it is started",
```
Comment (cboddy) [inline src/peergos/server/Start.java]: miinor: unclear

File: src/peergos/server/Start.java [hunk 6]
```diff
@@ -86,9 +90,87 @@
...
+                }
+            },
+            Collections.emptyList()
+    );
+
     public static final Command LOCAL = new Command("local",
             "Start an ephemeral Peergos Server and CoreNode server",
             args -> {
+                BOOTSTRAP.main(args);
                 args.setIfAbsent("domain", "localhost");
```
Comment (cboddy) [inline src/peergos/server/Start.java]: TODO: stop modifying ``args`` and passing it to other  things that modify it.

File: src/peergos/server/Start.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/server/Start.java]: done

File: src/peergos/server/Start.java (hunk 6 above)
Comment (ianopolous) [reply src/peergos/server/Start.java]: a future pr

File: src/peergos/server/Start.java (hunk 5 above)
Comment (ianopolous) [reply src/peergos/server/Start.java]: done

**Comment by ianopolous** [comment]
I've decided I'll do the secret key file stuff in a separate PR and merge this.

**Pull request merged**